HTML dialog element

estelle opened the pull request:
Added a medium example
changed simple to basic
Move a11y to correct location
rewrote a lot of content now that dialog is well-supported.

estelle committed:
HTML dialog element

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -51,13 +46,70 @@ The following will render a non-modal, or modal-less, dialog. The "OK" button al
+### Return value example
 
-This example opens a modal dialog when the "Show the dialog" button is activated. The dialog contains a form with a {{HTMLElement("select")}} and two {{HTMLElement("button")}} elements which default to `type="submit"`. Updating the value of the `<select>` updates the value of the "confirm" button. This value is the [`returnValue`](/en-US/docs/Web/API/HTMLDialogElement/returnValue). If the dialog  ...
+TThe value of the submitting button whose activation submits a form within a dialog is the [`returnValue`](/en-US/docs/Web/API/HTMLDialogElement/returnValue) of the dialog.
+
+This example opens a modal dialog when the "Show the dialog" button is activated. The dialog contains a form with a {{HTMLElement("select")}} and two {{HTMLElement("button")}} elements which default to `type="submit"`. An eventlistener updates the value of the "confirm" button when the `<select>` changes. This is the return value. If the dialog is closed with the <kbd>Esc</kbd> key, there is no r ...
+
+When the dialog is closed, the return value is displayed under the "Show the dialog" button.
```
- Can we start the description with something like "This example demonstrates how to..."

```suggestion
This example demonstrates how to open a modal dialog by using a form. The value of the button that submits the form within the `<dialog>` element is the [`returnValue`](/en-US/docs/Web/API/HTMLDialogElement/returnValue) of the dialog.

This example opens a modal dialog when the "Show the dialog" button is activated. The dialog contains a form with a {{HTMLElement("select")}} and two {{HTMLElement("button")}} elements, which default to `type="submit"`. An eventlistener updates the value of the "Confirm" button when the select option changes. This is the return value. If the dialog is closed by pressing the <kbd>Esc</kbd> key, there is no return value.

When the dialog is closed, the return value is displayed under the "Show the dialog" button.
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -51,13 +46,70 @@ The following will render a non-modal, or modal-less, dialog. The "OK" button al
+When a modal dialog is opened, it is opened over the top of any other dialogs that might be present. Everything outside the modal dialog are inert with interactions outside the dialog being blocked. You'll note when the dialog is opened, With the exception of the dialog, interacting with the document is not possible; the "Show the dialog" button is mostly obfuscated by the almost opaque backdrop  ...
 
-### Advanced example
+### Return value example
```
Something like this?

```suggestion
### Handling the return value from the dialog
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -126,8 +178,10 @@ confirmBtn.addEventListener("click", (event) => {
 
 {{EmbedLiveSample("Advanced_example", "100%", 300)}}
```
I can't preview and test the example because it is pointing to the previous section
```suggestion
{{EmbedLiveSample("Handling the return value from the dialog", "100%", 300)}}
```
..or whatever the example title is updated to

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -126,8 +178,10 @@ confirmBtn.addEventListener("click", (event) => {
 {{EmbedLiveSample("Advanced_example", "100%", 300)}}
 
-This modal dialog can be closed three ways. For keyboard users, modal dialogs can be closed with the <kbd>Esc</kbd> key. In this example, the "Cancel" button closes the dialog via the `dialog` form method and the "Confirm" closes the dialog via the {{domxref("HTMLDialogElement.close()")}} method.
+This example demonstrated three methods of closing modal dialogs; this dialog can be closed via an in-dialog form submitted using the `dialog` methode methods, as we also saw in the [HTML-only example](#html-only-example). This dialog can also be closed with the <kbd>Esc</kbd> key and via the {domxref("HTMLDialogElement.close()")}} method, also demonstrated in the [modal example](#modal-example). ...
```
```suggestion
This example demonstrates the following three methods of closing modal dialogs:
- By submitting the form within the dialog form using the `dialog` method (as seen in the [HTML-only example](#creating_an_html-only_modeless_dialog)).
- By pressing the <kbd>Esc</kbd> key.
- By calling the {domxref("HTMLDialogElement.close()")}} method (as seen in the [modal example](#creating_a_modal_dialog).
In this example, the "Cancel" button closes the dialog via the `dialog` form method and the "Confirm" button closes the dialog via the {{domxref("HTMLDialogElement.close()")}} method.
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -126,8 +178,10 @@ confirmBtn.addEventListener("click", (event) => {
-This modal dialog can be closed three ways. For keyboard users, modal dialogs can be closed with the <kbd>Esc</kbd> key. In this example, the "Cancel" button closes the dialog via the `dialog` form method and the "Confirm" closes the dialog via the {{domxref("HTMLDialogElement.close()")}} method.
+This example demonstrated three methods of closing modal dialogs; this dialog can be closed via an in-dialog form submitted using the `dialog` methode methods, as we also saw in the [HTML-only example](#html-only-example). This dialog can also be closed with the <kbd>Esc</kbd> key and via the {domxref("HTMLDialogElement.close()")}} method, also demonstrated in the [modal example](#modal-example). ...
+
 The "Cancel" button includes a [`formmethod="dialog"`](/en-US/docs/Web/HTML/Element/input/submit#formmethod), which overrides the {{HTMLElement("form")}}'s default {{HTTPMethod("GET")}} [`method`](/en-US/docs/Web/HTML/Element/form#method). When a form's method is [`dialog`](#usage_notes), the state of the form is saved, not submitted, and the dialog gets closed.
```
```suggestion
The "Cancel" button includes the [`formmethod="dialog"`](/en-US/docs/Web/HTML/Element/input/submit#formmethod) attribute, which overrides the {{HTMLElement("form")}}'s default {{HTTPMethod("GET")}} [`method`](/en-US/docs/Web/HTML/Element/form#method). When a form's method is [`dialog`](#usage_notes), the state of the form is saved but not submitted, and the dialog gets closed.
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -179,6 +233,18 @@ It is important to provide a closing mechanism within every `dialog` element. Th
 
+## Accessibility considerations
+
+While dialogs can be created using other elements, the native `<dialog>` element provides usability and accessibility features that must be replicated if using other elements for a similar purpose. If creating a custom dialog implementation, ensure all expected default behaviors are supported and proper labeling recommendations are followed.
```
```suggestion
While dialogs can be created using other elements, the native `<dialog>` element provides usability and accessibility features that must be replicated if you use other elements for a similar purpose. If you're creating a custom dialog implementation, ensure that all expected default behaviors are supported and proper labeling recommendations are followed.
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -179,6 +233,18 @@ It is important to provide a closing mechanism within every `dialog` element. Th
+
+While dialogs can be created using other elements, the native `<dialog>` element provides usability and accessibility features that must be replicated if using other elements for a similar purpose. If creating a custom dialog implementation, ensure all expected default behaviors are supported and proper labeling recommendations are followed.
+
+When implementing a dialog, it is important to consider the most appropriate place to set user focus. When using {{domxref("HTMLDialogElement.showModal()")}} to open a `<dialog>`, focus is set on the first nested focusable element. Explicitly indicating the initial focus placement by use of the [autofocus](/en-US/docs/Web/HTML/Global_attributes/autofocus) attribute will help ensure initial focus  ...
```
```suggestion
When implementing a dialog, it is important to consider the most appropriate place to set user focus. When using {{domxref("HTMLDialogElement.showModal()")}} to open a `<dialog>`, focus is set on the first nested focusable element. Explicitly indicating the initial focus placement by using the [`autofocus`](/en-US/docs/Web/HTML/Global_attributes/autofocus) attribute will help ensure initial focus is set on the element deemed the best initial focus placement for any particular dialog. When in doubt, as it may not always be known where initial focus could be set within a dialog, particularly for instances where a dialog's content is dynamically rendered when invoked, the `<dialog>` element itself may provide the best initial focus placement.
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -179,6 +233,18 @@ It is important to provide a closing mechanism within every `dialog` element. Th
+
+When implementing a dialog, it is important to consider the most appropriate place to set user focus. When using {{domxref("HTMLDialogElement.showModal()")}} to open a `<dialog>`, focus is set on the first nested focusable element. Explicitly indicating the initial focus placement by use of the [autofocus](/en-US/docs/Web/HTML/Global_attributes/autofocus) attribute will help ensure initial focus  ...
+
+Ensure a mechanism is provided to allow users to close a dialog. The most robust way to ensure all users can close a dialog is to include an explicit button to do so, sucha as a confirmation, cancel or close button.
```
```suggestion
Ensure a mechanism is provided to allow users to close the dialog. The most robust way to ensure that all users can close the dialog is to include an explicit button to do so, such as a confirmation, cancellation, or close button.
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -179,6 +233,18 @@ It is important to provide a closing mechanism within every `dialog` element. Th
+
+Ensure a mechanism is provided to allow users to close a dialog. The most robust way to ensure all users can close a dialog is to include an explicit button to do so, sucha as a confirmation, cancel or close button.
+
+By default, a `<dialog>` invoked by the `showModal()` method will allow for its dismissal by the <kbd>Escape</kbd>. A non-modal dialog does not dismiss via the <kbd>Escape</kbd> key by default, and depending on what the non-modal dialog represents, it may not be desired for this behavior. Keyboard users expect the <kbd>Escape</kbd> key to close modal dialogs; ensure this behavior is implemented a ...
```
```suggestion
By default, a dialog invoked by the `showModal()` method can be dismissed by pressing the <kbd>Esc</kbd> key. A non-modal dialog does not dismiss via the <kbd>Esc</kbd> key by default, and depending on what the non-modal dialog represents, it may not be desired for this behavior. Keyboard users expect the <kbd>Esc</kbd> key to close modal dialogs; ensure that this behavior is implemented and maintained. If multiple modal dialogs are open, pressing the <kbd>Esc</kbd> key should close only the last shown dialog. When using `<dialog>`, this behavior is provided by the browser.
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -179,6 +233,18 @@ It is important to provide a closing mechanism within every `dialog` element. Th
+
+By default, a `<dialog>` invoked by the `showModal()` method will allow for its dismissal by the <kbd>Escape</kbd>. A non-modal dialog does not dismiss via the <kbd>Escape</kbd> key by default, and depending on what the non-modal dialog represents, it may not be desired for this behavior. Keyboard users expect the <kbd>Escape</kbd> key to close modal dialogs; ensure this behavior is implemented a ...
+
+The `<dialog>` element is exposed by browsers similarly to custom dialogs using the ARIA [role="dialog"](/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role) attribute. `<dialog>` elements invoked by the `showModal()` method will have an implicit [aria-modal="true"](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-modal), whereas `<dialog>` elements invoked by the `show()` method, or rendered  ...
```
```suggestion
The `<dialog>` element is exposed by browsers in a a manner similar to custom dialogs that use the ARIA [role="dialog"](/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role) attribute. `<dialog>` elements invoked by the `showModal()` method implicitly have [aria-modal="true"](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-modal), whereas `<dialog>` elements invoked by the `show()` method or displayed using the `open` attribute or by changing the default `display` of a `<dialog>` are exposed as `[aria-modal="false"]`. When implementing modal dialogs, everything other than the `<dialog>` and its contents should be rendered inert using the [`inert`](/en-US/docs/Web/HTML/Global_attributes/inert) attribute. When using `<dialog>` along with the `HTMLDialogElement.showModal()` method, this behavior is provided by the browser.
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -193,4 +259,3 @@ It is important to provide a closing mechanism within every `dialog` element. Th
 - The {{domxref("HTMLDialogElement/cancel_event", "cancel")}} event
 - [HTML forms guide](/en-US/docs/Learn/Forms).
 - The {{cssxref("::backdrop")}} pseudo-element
```
```suggestion
- [Web forms](/en-US/docs/Learn/Forms) in the Learn area
```

dipikabh reviewed (changes requested):
Thanks for these updates @estelle 🙌. I've added my edits. I could not view the last example.

I looked into the use of the terms "non-modal" vs "modalless" vs "modeless":

On MDN, we have 13 occurrences of "non-modal" and 3 of "modeless" (single 'l'). So we can choose to stick to "non-modal". But "modeless" seems to be pretty common overall.
"modalless" is not recognized. (One of the references: https://en.wikipedia.org/wiki/Modal_window)

hamishwillee commented:
> On MDN, we have 13 occurrences of "non-modal" and 3 of "modeless" (single 'l'). So we can choose to stick to "non-modal". But "modeless" seems to be pretty common overall.

FWIW I have never heard the term modeless used in a dialog context. Modal and non-modal are correct.

estelle committed:
Apply suggestions from code review

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

estelle commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -19,26 +21,19 @@ This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attrib
 ## Examples
 
-### Simple example
+### HTML-only example
```
There are issues with HTML-only examples. This example is "not advised".  An HTML-only dialog must be loaded open and once closed, can not be re-opened (that requires JS), so want a non-promotional header.

dipikabh replied:
Makes sense. I'll make a suggestion to reflect this through the title.

estelle committed:
Update files/en-us/web/html/element/dialog/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

estelle committed:
Update files/en-us/web/html/element/dialog/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

estelle committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

estelle committed:
Many edits

estelle commented:
thanks @dipikabh for the thorough edits. I included them all except one (the title of the HTML only thing). I updated all modeless to "non-modal" per hamish (and because i still can get my head around changing the a to an e.)
Some suggestions made me realize that my explanations weren't good enough, so i updated the return value section a lot.

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -189,8 +269,10 @@ It is important to provide a closing mechanism within every `dialog` element. Th
-- [HTML forms guide](/en-US/docs/Learn/Forms).
-- The {{cssxref("::backdrop")}} pseudo-element
-- [dialog-polyfill](https://github.com/GoogleChrome/dialog-polyfill)
+- The {{domxref("HTMLDialogElement/open", "open")}} property
+- The {{CSSXref("::backdrop")}} pseudo-element
+- The [`inert`](/en-US/docs/Web/HTML/Global_attributes/inert) attribute
```
- We can drop the "The"s
- Maybe we can qualify the property and attribute
- Also changing the order to group technologies
```suggestion
- {{domxref("HTMLDialogElement/open", "open")}} property of the `HTMLDialogElement` interface
- [`inert`](/en-US/docs/Web/HTML/Global_attributes/inert) global attribute for HTML elements
- {{CSSXref("::backdrop")}} CSS pseudo-element
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -189,8 +269,10 @@ It is important to provide a closing mechanism within every `dialog` element. Th
 
 ## See also
 
+- The {{domxref("HTMLDialogElement/")}} Interface
 - The {{domxref("HTMLDialogElement/close_event", "close")}} event
 - The {{domxref("HTMLDialogElement/cancel_event", "cancel")}} event
```
```suggestion
- {{domxref("HTMLDialogElement")}} interface
- {{domxref("HTMLDialogElement/close_event", "close")}} event
- {{domxref("HTMLDialogElement/cancel_event", "cancel")}} event
```

dipikabh commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -16,29 +20,23 @@ This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attrib
 ## Examples
 
-### Simple example
+### HTML-only example
```
```suggestion
### Caveats of creating a non-modal dialog using only HTML
```

dipikabh reviewed (changes requested):
Thanks for the updates! I agree - better to go with "non-modal". I've added a suggestion for rewording the HTML-only example to reflect that it is not recommended. And I spotted a few typos from my previous edits (sorry!).

estelle committed:
Apply suggestions from code review

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

estelle committed:
final edits?

estelle reviewed (commented):
Thanks @dipikabh. Suggestions incorporated!

dipikabh reviewed (changes requested):
Thank you for the updates!

dipikabh commented:
Thank you for the updates!
